pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,282 @@
+		return nil, err
+	}
+	aggregatedGroups, err := d.findAggregatedGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resourceToGroupVersions := make(map[string][]schema.GroupVersion)
+	_, resourceLists, err := d.discoveryClient.ServerGroupsAndResources()
+	if err != nil {
+		return nil, err
+	}
+	for _, resourceList := range resourceLists {
+		gv, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+		if err != nil {
+			klog.Errorf("cannot parse group version %s, ignored", resourceList.GroupVersion)
+			continue
+		}
+		if customGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's a custom group", gv.Group)
+			continue
+		}
+		if aggregatedGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's an aggregated group", gv.Group)
+			continue
+		}
+		for _, r := range resourceList.APIResources {
+			// ignore subresources
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			if blackListResources.Has(r.Name) {
+				continue
+			}
+			// ignore resources that cannot be listed and updated
+			if !sets.NewString(r.Verbs...).HasAll("list", "update") {
+				continue
+			}
+			gvs := resourceToGroupVersions[r.Name]
+			gvs = append(gvs, gv)
+			resourceToGroupVersions[r.Name] = gvs
+		}
+	}
+
+	ret := []schema.GroupVersionResource{}
+	for resource, groupVersions := range resourceToGroupVersions {
+		if len(groupVersions) == 1 {
+			continue
+		}
+		ret = append(ret, groupVersions[0].WithResource(resource))
+	}
+	return ret, nil
+}
+
+func (m *migrator) migrateOneItem(ctx context.Context, resource schema.GroupVersionResource, item *unstructured.Unstructured) error {
+	namespace, err := metadataAccessor.Namespace(item)
+	if err != nil {
+		return err
+	}
+	name, err := metadataAccessor.Name(item)
+	if err != nil {
+		return err
+	}
+	getBeforePut := false
+	for {
+		getBeforePut, err = m.try(ctx, resource, namespace, name, item, getBeforePut)
+		if err == nil || errors.IsNotFound(err) {
+			return nil
+		}
+		if canRetry(err) {
+			seconds, delay := errors.SuggestsClientDelay(err)
+			switch {
+			case delay && len(namespace) > 0:
+				klog.Warningf("migration of %s, in the %s namespace, will be retried after a %ds delay: %v", name, namespace, seconds, err)
+				time.Sleep(time.Duration(seconds) * time.Second)
+			case delay:
+				klog.Warningf("migration of %s will be retried after a %ds delay: %v", name, seconds, err)
+				time.Sleep(time.Duration(seconds) * time.Second)
+			case !delay && len(namespace) > 0:
+				klog.Warningf("migration of %s, in the %s namespace, will be retried: %v", name, namespace, err)
+			default:
+				klog.Warningf("migration of %s will be retried: %v", name, err)
+			}
+			continue
+		}
+		// error is not retriable
+		return fmt.Errorf("can not retry: %+v", err)
+	}
+}
+
+func (m *migrator) try(ctx context.Context, resource schema.GroupVersionResource, namespace, name string, item *unstructured.Unstructured, get bool) (bool, error) {
+	var err error
+	if get {
```

Comment:
Do we really need that? Could we always "get before put"?

I can image it's an extra load but at this point nothing is running on the cluster so MicroShift gets all the resources for itself.
On the other hand, what's the chance of the conflict if nothing is running... (but I'd rather have a "get before put" than end up with conflict for some weird reason that retries would remediate)

Reply:
It's more there for a safety concern, it won't execute unless there's a conflict, so if we do decide to run this during normal operation or what may happen, this would cover us.

Reply:
Do you see a benefit in running this version of storage migration code under "normal" operations, instead of running the standard controller?

Reply:
The one benefit I can think of is having an easier user experience path for re-running a migration instead of having to install the storage migration controller. In the event you would want to execute a migration after a resource conflict has been resolved manually, having a cockpit plugin button or running a `microshift admin migrate` command I think would be for a better flow. This assumes it would be the admin who would install it and not us.

Reply:
If the service won't start, there's not going to be a way to run any code that modifies the data, right? The user is going to have to roll back and do something to the system using the old code versions.

What sorts of "conflicts" are we talking about?

Reply:
Yeah you're right, this is over preparing for a future where a user might want to re-run a migration that takes a long time during normal operation. Removed it.